ui/packages/app/src/api.ts

```diff
@@ -17,10 +17,22 @@ export const fn = async (
 	session_hash: string,
 	api_endpoint: string,
 	action: string,
+	js: string | false,
 	data: Record<string, unknown>,
 	queue: boolean,
 	queue_callback: (pos: number | null, is_initial?: boolean) => void
 ) => {
+	if (js !== false) {
+		console.log(js);
+		try {
+			var jsfn = eval(js);
```

Comment:
We should do this up front when we are parsing the config rather than doing ti everytime we make a call. The sooner we can treat this as a normal function the easier our lives will be.

Additionally, [`new Function()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function) is preferable to `eval` as it is faster and has safer scoping rules.

Something like this shoudl give the same result

```js
const jsfn = new Function(`return ${js}`)();
```

Reply:
Done. Instead of combining input data in one list called data and output data in one arg called state, I spread the args to match the style used in python - the output state values come right after input arguments.